index.html

```diff
@@ -3981,6 +4218,124 @@ <h3>
+              country B, a developer should request only "country" first and
+              decide if "region" is needed.
+            </p>
+            <p>
+              A developer can call {{requestShippingAddress()}} repeatedly.
+              Each call returns a promise that resolves to a {{PaymentAddress}}
+              with all previously requested address fields populated, subject to
+              any applicable <a data-lt="shipping-redact-list">redact list</a>.
+            </p>
+            <pre class="example js" title=
+            "how to use `requestShippingAddress()` correctly.">
+              // ✅ Good - Request additional address parts to recompute
+              // shipping cost. The final response.shippingAddress will also
+              // only contain "country" and "postalCode".
+              request.onshippingaddresschange = (ev) =&gt; {
+                ev.updateWith(new Promise((resolver) =&gt; {
+                  ev.requestShippingAddress(["country", "postalCode"])
+                      .then(address =&gt; {
+                        resolver(getNewDetails(address));
+                      });
+                  }));
+              };
+
+              // ✅ Good - Request additional addresss parts to be included in
+              // the final response.shippingAddress.
+              request.onshippingaddresschange = (ev) =&gt; {
+                // Compute shipping cost based on initial coarse address.
+                ev.updateWith(new Promise((resolver) =&gt; {
+                  // Discard the returned promise because the full address is not
+                  // needed until the final PaymentResponse. Empty addressParts
+                  // means all parts are requested.
+                  ev.requestShippingAddress(/* addressParts= */[]);
+
+                  // Compute shipping cost using initial partial address.
+                  resolver(computeShippingCost(ev.target.shippingAddress)));
+                });
+              }
+
+              // ❌ Bad - this won't work!
+              request.onshippingaddresschange = (ev) =&gt; {
+                ev.updateWith(getPromiseForNewDetails(ev.target.shippingAddress));
+
+                // This will throw InvalidStateError since not inside updateWith().
+                ev.requestShippingAddress([]);
+              };
+
+              // ❌ Bad - this won't work!
+              request.onshippingaddresschange = (ev) =&gt; {
+                // This will throw InvalidStateError since not inside updateWith().
+                ev.requestShippingAddress([]);
+
+                ev.updateWith(getPromiseForNewDetails(ev.target.shippingAddress));
+              };
+            </pre>
+          </aside>
+          <p>
+            The <a>requestShippingAddress(|addressParts|)</a> method MUST act as
+            follows:
+          </p>
+          <ol class="algorithm">
+            <li>Let |event:PaymentRequestUpdateEvent| be this
```

Comment:
the `requestShippingAddress()` and `requestBillingAddress()` steps are almost identical. I wonder if we should generalize them to a "request address" algorithm and pass in a /type/ and then "if type is 'billing, do x. If it's 'shipping', do y"?

Reply:
I considered that, but unfortunately they are just different enough I think a combined algorithm will have a lot of "if "billing", do x, else if "shipping", do y", which I think is less readable.

After Shipping Delegation [1] ships, I think we will be able to refactor all shipping / contact information algorithms to align with billing address, as all information will come from the payment handler. Can I defer this refactor until then?

[1] https://github.com/w3c/payment-handler/pull/349

Reply:
I'm ok to defer.